### ISSUE TITLE
Create a text file with hashes signed with GPG for Linux releases

### ISSUE THREAD

**Issue by Lew-Rockwell-Fan** [webtorrent/webtorrent-desktop#1769]
I just downloaded the new deb.  If there are any published hashes, let alone signed hashes, I didn't  see them. It's not that hard, guys. In 2019, it's pretty much a basic security procedure to hash your final products with at least a sha256 sum and publish a gpg signed text file with all your hashes.  Learning gpg sufficiently from square 0 shouldn't take more than 30 minutes. Establishing a public key and publishing it, maybe 10. Hashing the files and signing the list of hashes maybe another 10. At most.

**Comment by hicom150** [comment]
Could you please elaborate a bit more what steps should we follow to create this gpg signed text file?
Do you have any interesting link to documentation for a newbie like me? 😅

**Comment by Lew-Rockwell-Fan** [comment]
Sure. Get a hash for each file:
In bash (& probably any shell) on a gnu/linux (the '# ' is my prompt):
               
# sha512sum testfilea > hashes.txt
# sha512sum testfileb >> hashes.txt
# sha512sum testfilec >> hashes.txt

To check that:
# cat hashes.txt
77e52614a141cef19a76de6193e9c845b8cffffecf4973653e5d7a5d121450fab6142890e0f794715722ba4257ee28cb0f4e3966f18e11c9e62fa6babf0a748f  testfilea
beeea721aa75335fada24353bbdd08f3bca1930f3e03073dbb4b55e7147f4f72f1f02f49cd3a1440619878a483f70432dc972a20b1b0c49f9e4344ed91cf494a  testfileb
a730478525b2582111f337d726a93649239c5d0fd2dfaa004c04448d1310532cb05051e20c95ad7d5f48c0fd4b60ea2a29fd0b436d8ae38a449192c66629d8f8  testfilec

It's been a long time since I messed with Windows, but https://bhoover.com/how-to-verify-checksum-windows/
says that as of a year ago you could get a sha256sum with

certutil -hashfile SHA256 [presumably with a filename argument here]

I'd assume you can substitute 512 for 256, but realistically, either is adequate.

To sign the resulting text file, you need either PGP or GPG. GPG is cross platform & FOSS. PGP is proprietary (but I believe it IS open source, which contrary to popular perception, is a separate issue) & I don't know if they have a linux version. A good place to read on GPG is
https://gnupg.org/
or, if you run a 'nix, there is a man who will explain it all: 
man gpg
There is a gui called seahorse (I believe there are others too) or you can do everything from the command line. Here's another decent how-to:
https://www.digitalocean.com/community/tutorials/how-to-use-gpg-to-encrypt-and-sign-messages
I'm not current on the Windows versions but Metager is your friend. Because the G-monster is nobody's friend. 
Basically, you do 3 things:
- generate the text file with hashes for all files you publish
- establish a public/private keypair (see the gpg docs)
-either produce a signed and encrypted version of the hashes.txt file OR produce a separate signature file it is to be verified with.  The first approach is simpler but the second does make the hashes available (albeit insecurely) to people who won;t take the extra seconds to use gpg in addition.
- publish that file or files, your gpg pubic key, on your site and publish the key to public  keyservers. MIT's is widely trusted. Also Ubuntu runs one.

**Comment by feross** [comment]
We already sign the Mac and Windows binaries. To my knowledge, there's no established standard for signing binaries across Linux flavors, which is unfortunate. Creating a text file with hashes is a reasonable compromise but this is low priority for us at the moment. If you'd like to see it implemented, please consider submitting a PR.

**Comment by Lew-Rockwell-Fan** [comment]
_"there's no established standard for signing binaries across Linux flavors"_ There may not be a formally adopted standard documented in a publication, like, for example, the Desktop standards, or the standards for what constitutes posix, but what I outlined for you is the de facto standard and has been for a long time. If your project is accepted into a distro's repo, they'll do their own version of it, to work transparently with whatever package management system they use, although that's still probably the same thing under the hood.  Otherwise, a pubic asymmetric gpp key & a signed file with hash values is how it's done. Anybody who doesn't, doesn't take security seriously. That's OK if your product is a hundred line bash script, but not for anything less transparent. As for me doing it, I'm not in a position to hash your files. How would I know I had good copies that hadn't been tampered with? That's the whole point of doing it after all.  The same applies in spades for signing it. But don't bother on my account. I'm using a different solution now. I only dropped in because I noticed I had an unread notice when I came to github for other reasons.
